File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/resource/ResourceErrorReason.java
package org.maplibre.nativeffi.resource;

/** Native resource error reason copied from events or resource responses. */
public final class ResourceErrorReason {
  public static final ResourceErrorReason NONE = new ResourceErrorReason(0);
  public static final ResourceErrorReason NOT_FOUND = new ResourceErrorReason(1);
  public static final ResourceErrorReason SERVER = new ResourceErrorReason(2);
  public static final ResourceErrorReason CONNECTION = new ResourceErrorReason(3);
  public static final ResourceErrorReason RATE_LIMIT = new ResourceErrorReason(4);
  public static final ResourceErrorReason OTHER = new ResourceErrorReason(5);

  private final int rawValue;
  private final String name;

  public ResourceErrorReason(int rawValue) {
    this.rawValue = rawValue;
    this.name =
        switch (rawValue) {
          case 0 -> "NONE";
          case 1 -> "NOT_FOUND";
          case 2 -> "SERVER";
          case 3 -> "CONNECTION";
          case 4 -> "RATE_LIMIT";
          case 5 -> "OTHER";
          default -> "UNKNOWN(" + rawValue + ")";
        };
  }

  public int rawValue() {
    return rawValue;
  }

  @Override
  public boolean equals(Object other) {
    return other instanceof ResourceErrorReason that && rawValue == that.rawValue;
  }

  @Override
  public int hashCode() {
    return Integer.hashCode(rawValue);
  }

  @Override
  public String toString() {
    return name;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/resource/ResourceErrorReason.kt
package org.maplibre.nativeffi.resource

import kotlin.jvm.JvmInline

/**
 * Resource error reason copied from native events.
 *
 * This is an open domain: MapLibre Native may report a value that has no named constant here, so a
 * `when` over this type needs an `else` branch. Unknown values are preserved as their raw
 * [nativeValue] rather than collapsed to a known constant.
 */
@JvmInline
public value class ResourceErrorReason(public val nativeValue: Int) {
  public companion object {
    public val NONE: ResourceErrorReason = ResourceErrorReason(0)
    public val NOT_FOUND: ResourceErrorReason = ResourceErrorReason(1)
    public val SERVER: ResourceErrorReason = ResourceErrorReason(2)
    public val CONNECTION: ResourceErrorReason = ResourceErrorReason(3)
    public val RATE_LIMIT: ResourceErrorReason = ResourceErrorReason(4)
    public val OTHER: ResourceErrorReason = ResourceErrorReason(5)

    internal fun fromNative(nativeValue: UInt): ResourceErrorReason =
      fromNative(nativeValue.toInt())

    internal fun fromNative(nativeValue: Int): ResourceErrorReason =
      ResourceErrorReason(nativeValue)
  }

  internal val isKnown: Boolean
    get() =
      this == NONE ||
        this == NOT_FOUND ||
        this == SERVER ||
        this == CONNECTION ||
        this == RATE_LIMIT ||
        this == OTHER
}
